In [1]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv("Online Retail (3).xlsx - Online Retail.csv")
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  
541906  12/9/11 12:50       4.15     12680.0          France  
541907  12/9/11 12:50       4.15     12680.0          France  
541908  12/9/11 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

creating copy of data frame

In [4]:
df1 = df.copy()
df1

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  
541906  12/9/11 12:50       4.15     12680.0          France  
541907  12/9/11 12:50       4.15     12680.0          France  
541908  12/9/11 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

Analysng given data

In [5]:
df1.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [7]:
df1.shape

(541909, 8)

In [11]:
df1.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

thus there is null values in Description and CustomerID column

In [12]:
df1.head(20)

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5     536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6     536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7     536366     22633               HAND WARMER UNION JACK         6   
8     536366     22632            HAND WARMER RED POLKA DOT         6   
9     536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   
10    536367     22745           POPPY'S PLAYHOUSE BEDROOM          6   
11    536367     22748            POPPY'S PLAYHOUSE KITCHEN         6   
12    536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL         8   
13    536367     22310              IVORY KNITTED MUG COSY          6   
14    536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS         6   
15    536367     22623        BOX OF VINTAGE JIGSAW BLOCKS          3   
16    536367     22622       BOX OF VINTAGE ALPHABET BLOCKS         2   
17    536367     21754             HOME BUILDING BLOCK WORD         3   
18    536367     21755             LOVE BUILDING BLOCK WORD         3   
19    536367     21777          RECIPE BOX WITH METAL HEART         4   

     InvoiceDate  UnitPrice  CustomerID         Country  
0   12/1/10 8:26       2.55     17850.0  United Kingdom  
1   12/1/10 8:26       3.39     17850.0  United Kingdom  
2   12/1/10 8:26       2.75     17850.0  United Kingdom  
3   12/1/10 8:26       3.39     17850.0  United Kingdom  
4   12/1/10 8:26       3.39     17850.0  United Kingdom  
5   12/1/10 8:26       7.65     17850.0  United Kingdom  
6   12/1/10 8:26       4.25     17850.0  United Kingdom  
7   12/1/10 8:28       1.85     17850.0  United Kingdom  
8   12/1/10 8:28       1.85     17850.0  United Kingdom  
9   12/1/10 8:34       1.69     13047.0  United Kingdom  
10  12/1/10 8:34       2.10     13047.0  United Kingdom  
11  12/1/10 8:34       2.10     13047.0  United Kingdom  
12  12/1/10 8:34       3.75     13047.0  United Kingdom  
13  12/1/10 8:34       1.65     13047.0  United Kingdom  
14  12/1/10 8:34       4.25     13047.0  United Kingdom  
15  12/1/10 8:34       4.95     13047.0  United Kingdom  
16  12/1/10 8:34       9.95     13047.0  United Kingdom  
17  12/1/10 8:34       5.95     13047.0  United Kingdom  
18  12/1/10 8:34       5.95     13047.0  United Kingdom  
19  12/1/10 8:34       7.95     13047.0  United Kingdom

In [13]:
df1.tail(20)

InvoiceNo StockCode                          Description  Quantity  \
541889    581585     22466       FAIRY TALE COTTAGE NIGHT LIGHT        12   
541890    581586     22061  LARGE CAKE STAND  HANGING STRAWBERY         8   
541891    581586     23275     SET OF 3 HANGING OWLS OLLIE BEAK        24   
541892    581586     21217        RED RETROSPOT ROUND CAKE TINS        24   
541893    581586     20685                DOORMAT RED RETROSPOT        10   
541894    581587     22631             CIRCUS PARADE LUNCH BOX         12   
541895    581587     22556       PLASTERS IN TIN CIRCUS PARADE         12   
541896    581587     22555            PLASTERS IN TIN STRONGMAN        12   
541897    581587     22728            ALARM CLOCK BAKELIKE PINK         4   
541898    581587     22727            ALARM CLOCK BAKELIKE RED          4   
541899    581587     22726           ALARM CLOCK BAKELIKE GREEN         4   
541900    581587     22730           ALARM CLOCK BAKELIKE IVORY         4   
541901    581587     22367      CHILDRENS APRON SPACEBOY DESIGN         8   
541902    581587     22629                  SPACEBOY LUNCH BOX         12   
541903    581587     23256          CHILDRENS CUTLERY SPACEBOY          4   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID         Country  
541889  12/9/11 12:31       1.95     15804.0  United Kingdom  
541890  12/9/11 12:49       2.95     13113.0  United Kingdom  
541891  12/9/11 12:49       1.25     13113.0  United Kingdom  
541892  12/9/11 12:49       8.95     13113.0  United Kingdom  
541893  12/9/11 12:49       7.08     13113.0  United Kingdom  
541894  12/9/11 12:50       1.95     12680.0          France  
541895  12/9/11 12:50       1.65     12680.0          France  
541896  12/9/11 12:50       1.65     12680.0          France  
541897  12/9/11 12:50       3.75     12680.0          France  
541898  12/9/11 12:50       3.75     12680.0          France  
541899  12/9/11 12:50       3.75     12680.0          France  
541900  12/9/11 12:50       3.75     12680.0          France  
541901  12/9/11 12:50       1.95     12680.0          France  
541902  12/9/11 12:50       1.95     12680.0          France  
541903  12/9/11 12:50       4.15     12680.0          France  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  
541906  12/9/11 12:50       4.15     12680.0          France  
541907  12/9/11 12:50       4.15     12680.0          France  
541908  12/9/11 12:50       4.95     12680.0          France

In [15]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


# Analysing invoice column

In [17]:
df1["InvoiceNo"].unique()

array(['536365', '536366', '536367', ..., '581585', '581586', '581587'],
      dtype=object)

In [18]:
df1["InvoiceNo"].nunique()

25900

thus there is 25900 unique values

In [19]:
df2 = df1[df1['InvoiceNo'].str.contains('C')]

In [20]:
df2

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

          InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/10 9:41      27.50     14527.0  United Kingdom  
154      12/1/10 9:49       4.65     15311.0  United Kingdom  
235     12/1/10 10:24       1.65     17548.0  United Kingdom  
236     12/1/10 10:24       0.29     17548.0  United Kingdom  
237     12/1/10 10:24       0.29     17548.0  United Kingdom  
...               ...        ...         ...             ...  
540449   12/9/11 9:57       0.83     14397.0  United Kingdom  
541541  12/9/11 10:28     224.69     15498.0  United Kingdom  
541715  12/9/11 11:57      10.95     15311.0  United Kingdom  
541716  12/9/11 11:58       1.25     17315.0  United Kingdom  
541717  12/9/11 11:58       1.25     17315.0  United Kingdom  

[9288 rows x 8 columns]

thus we have 9288 entry having C thus we have to remov that

In [21]:
df1["InvoiceNo"] = df1["InvoiceNo"].str.replace("C"," ") 
df1

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  
541906  12/9/11 12:50       4.15     12680.0          France  
541907  12/9/11 12:50       4.15     12680.0          France  
541908  12/9/11 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

checking is C is removed or not

In [23]:
df1 = df1[df1['InvoiceNo'].str.contains('C')]
df1

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

thus we have 0 entry in this column hece C is removed

# Analysing for column “CustomerID'

In [40]:
df1['CustomerID'].isnull().sum()


0

In [35]:
df1.describe()

Quantity  UnitPrice  CustomerID
count       0.0        0.0         0.0
mean        NaN        NaN         NaN
std         NaN        NaN         NaN
min         NaN        NaN         NaN
25%         NaN        NaN         NaN
50%         NaN        NaN         NaN
75%         NaN        NaN         NaN
max         NaN        NaN         NaN

In [37]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 0 entries
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   InvoiceNo    0 non-null      object 
 1   StockCode    0 non-null      object 
 2   Description  0 non-null      object 
 3   Quantity     0 non-null      int64  
 4   InvoiceDate  0 non-null      object 
 5   UnitPrice    0 non-null      float64
 6   CustomerID   0 non-null      float64
 7   Country      0 non-null      object 
dtypes: float64(2), int64(1), object(5)
memory usage: 0.0+ bytes


In [38]:
df1.isnull().sum()

InvoiceNo      0.0
StockCode      0.0
Description    0.0
Quantity       0.0
InvoiceDate    0.0
UnitPrice      0.0
CustomerID     0.0
Country        0.0
dtype: float64

In [43]:
df2 = df.copy()

In [44]:
df2

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  
541906  12/9/11 12:50       4.15     12680.0          France  
541907  12/9/11 12:50       4.15     12680.0          France  
541908  12/9/11 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [45]:
df2[df2['InvoiceNo'].str.contains('C')]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

          InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/10 9:41      27.50     14527.0  United Kingdom  
154      12/1/10 9:49       4.65     15311.0  United Kingdom  
235     12/1/10 10:24       1.65     17548.0  United Kingdom  
236     12/1/10 10:24       0.29     17548.0  United Kingdom  
237     12/1/10 10:24       0.29     17548.0  United Kingdom  
...               ...        ...         ...             ...  
540449   12/9/11 9:57       0.83     14397.0  United Kingdom  
541541  12/9/11 10:28     224.69     15498.0  United Kingdom  
541715  12/9/11 11:57      10.95     15311.0  United Kingdom  
541716  12/9/11 11:58       1.25     17315.0  United Kingdom  
541717  12/9/11 11:58       1.25     17315.0  United Kingdom  

[9288 rows x 8 columns]

In [46]:
df2["InvoiceNo"] = df2["InvoiceNo"].str.replace("C"," ") 
df2

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  
541906  12/9/11 12:50       4.15     12680.0          France  
541907  12/9/11 12:50       4.15     12680.0          France  
541908  12/9/11 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [47]:
df2[df2['InvoiceNo'].str.contains('C')]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

# Analysing for column “CustomerID'

In [49]:
df2['CustomerID'].isnull().sum()

135080

In [50]:
#Replacing nul values with 0.
df2["CustomerID"].fillna(0,inplace=True)

In [52]:
# checking whether null values replaced successfully or not
df2[df2["CustomerID"]==0]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

          InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/10 11:52       0.00         0.0  United Kingdom  
1443    12/1/10 14:32       2.51         0.0  United Kingdom  
1444    12/1/10 14:32       2.51         0.0  United Kingdom  
1445    12/1/10 14:32       0.85         0.0  United Kingdom  
1446    12/1/10 14:32       1.66         0.0  United Kingdom  
...               ...        ...         ...             ...  
541536  12/9/11 10:26       4.13         0.0  United Kingdom  
541537  12/9/11 10:26       4.13         0.0  United Kingdom  
541538  12/9/11 10:26       4.96         0.0  United Kingdom  
541539  12/9/11 10:26      10.79         0.0  United Kingdom  
541540  12/9/11 10:26    1714.17         0.0  United Kingdom  

[135080 rows x 8 columns]

In [54]:
# checking the numbers of null values in it.
df2["CustomerID"].isnull().sum()


0

thus null values is replaced by zero

# (4) Analysing for column “Quantity”

In [56]:
df2["Quantity"].isnull().sum()


0

In [58]:
df2["Quantity"].unique()

array([     6,      8,      2,     32,      3,      4,     24,     12,
           48,     18,     20,     36,     80,     64,     10,    120,
           96,     23,      5,      1,     -1,     50,     40,    100,
          192,    432,    144,    288,    -12,    -24,     16,      9,
          128,     25,     30,     28,      7,     56,     72,    200,
          600,    480,     -6,     14,     -2,     11,     33,     13,
           -4,     -5,     -7,     -3,     70,    252,     60,    216,
          384,    -10,     27,     15,     22,     19,     17,     21,
           34,     47,    108,     52,  -9360,    -38,     75,    270,
           42,    240,     90,    320,   1824,    204,     69,    -36,
         -192,   -144,    160,   2880,   1400,     39,    -48,    -50,
           26,   1440,     31,     82,     78,     97,     98,     35,
           57,    -20,    110,    -22,    -30,    -70,   -130,    -80,
         -120,    -40,    -25,    -14,    -15,    -69,   -140,   -320,
      

In [59]:
df2["Quantity"].nunique()

722

thus replacing negative by zero

In [61]:
df2.loc[df2["Quantity"]<0,'Quantity']=0

In [62]:
df2['Quantity'].unique()


array([    6,     8,     2,    32,     3,     4,    24,    12,    48,
          18,    20,    36,    80,    64,    10,   120,    96,    23,
           5,     1,     0,    50,    40,   100,   192,   432,   144,
         288,    16,     9,   128,    25,    30,    28,     7,    56,
          72,   200,   600,   480,    14,    11,    33,    13,    70,
         252,    60,   216,   384,    27,    15,    22,    19,    17,
          21,    34,    47,   108,    52,    75,   270,    42,   240,
          90,   320,  1824,   204,    69,   160,  2880,  1400,    39,
          26,  1440,    31,    82,    78,    97,    98,    35,    57,
         110,   720,   156,   324,    38,    37,    49,    95,    29,
          41,   402,   378,   150,   300,    54,   104,    67,   258,
          66,    44,    55,    46,    99,    61,   408,   972,   208,
        1008,  1000,  1488,   250,  1394,   400,    79,    84,   272,
         113,    45,   106,    68,   267,   115,    65,  1728,    53,
          76,   460,

thus sucessfully replaced by zero 

# Analysing description column

In [64]:
 # removing the special characters from the Description column
df2['Description'] = df2['Description'].str.replace(r'[^\w\s]', ' ', regex=True)

In [65]:
df2.loc[120:]

InvoiceNo StockCode                      Description  Quantity  \
120       536381    37444A  YELLOW BREAKFAST CUP AND SAUCER         1   
121       536381    37444C   PINK BREAKFAST CUP AND SAUCER          1   
122       536381     22086  PAPER CHAIN KIT 50 S CHRISTMAS          4   
123       536381     22083        PAPER CHAIN KIT RETROSPOT         1   
124       536381    84971S        SMALL HEART FLOWERS HOOK          6   
...          ...       ...                              ...       ...   
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID         Country  
120      12/1/10 9:41       2.95     15311.0  United Kingdom  
121      12/1/10 9:41       2.95     15311.0  United Kingdom  
122      12/1/10 9:41       2.95     15311.0  United Kingdom  
123      12/1/10 9:41       2.95     15311.0  United Kingdom  
124      12/1/10 9:41       0.85     15311.0  United Kingdom  
...               ...        ...         ...             ...  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  
541906  12/9/11 12:50       4.15     12680.0          France  
541907  12/9/11 12:50       4.15     12680.0          France  
541908  12/9/11 12:50       4.95     12680.0          France  

[541789 rows x 8 columns]

In [66]:
#checking is there any null value or not in it.
df2[df2['Description'].isnull()]


InvoiceNo StockCode Description  Quantity    InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/10 11:52        0.0   
1970      536545     21134         NaN         1  12/1/10 14:32        0.0   
1971      536546     22145         NaN         1  12/1/10 14:33        0.0   
1972      536547     37509         NaN         1  12/1/10 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/10 14:34        0.0   
...          ...       ...         ...       ...            ...        ...   
535322    581199     84581         NaN         0  12/7/11 18:26        0.0   
535326    581203     23406         NaN        15  12/7/11 18:31        0.0   
535332    581209     21620         NaN         6  12/7/11 18:35        0.0   
536981    581234     72817         NaN        27  12/8/11 10:33        0.0   
538554    581408     85175         NaN        20  12/8/11 14:06        0.0   

        CustomerID         Country  
622            0.0  United Kingdom  
1970           0.0  United Kingdom  
1971           0.0  United Kingdom  
1972           0.0  United Kingdom  
1987           0.0  United Kingdom  
...            ...             ...  
535322         0.0  United Kingdom  
535326         0.0  United Kingdom  
535332         0.0  United Kingdom  
536981         0.0  United Kingdom  
538554         0.0  United Kingdom  

[1454 rows x 8 columns]

there are 1454 null values

In [68]:
#replacing Nan value with ' '[whitespace]
df2["Description"].fillna(" ",inplace=True)

In [69]:
# crosschecking the Nan value
df2["Description"].isnull().sum()

0

# Analysng InvoiceDate columns

In [70]:
df2["InvoiceDate"]=pd.to_datetime(df2["InvoiceDate"]).dt.date

In [71]:
df2.head(20)

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART          6   
5     536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6     536365     21730    GLASS STAR FROSTED T LIGHT HOLDER         6   
7     536366     22633               HAND WARMER UNION JACK         6   
8     536366     22632            HAND WARMER RED POLKA DOT         6   
9     536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   
10    536367     22745           POPPY S PLAYHOUSE BEDROOM          6   
11    536367     22748            POPPY S PLAYHOUSE KITCHEN         6   
12    536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL         8   
13    536367     22310              IVORY KNITTED MUG COSY          6   
14    536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS         6   
15    536367     22623        BOX OF VINTAGE JIGSAW BLOCKS          3   
16    536367     22622       BOX OF VINTAGE ALPHABET BLOCKS         2   
17    536367     21754             HOME BUILDING BLOCK WORD         3   
18    536367     21755             LOVE BUILDING BLOCK WORD         3   
19    536367     21777          RECIPE BOX WITH METAL HEART         4   

   InvoiceDate  UnitPrice  CustomerID         Country  
0   2010-12-01       2.55     17850.0  United Kingdom  
1   2010-12-01       3.39     17850.0  United Kingdom  
2   2010-12-01       2.75     17850.0  United Kingdom  
3   2010-12-01       3.39     17850.0  United Kingdom  
4   2010-12-01       3.39     17850.0  United Kingdom  
5   2010-12-01       7.65     17850.0  United Kingdom  
6   2010-12-01       4.25     17850.0  United Kingdom  
7   2010-12-01       1.85     17850.0  United Kingdom  
8   2010-12-01       1.85     17850.0  United Kingdom  
9   2010-12-01       1.69     13047.0  United Kingdom  
10  2010-12-01       2.10     13047.0  United Kingdom  
11  2010-12-01       2.10     13047.0  United Kingdom  
12  2010-12-01       3.75     13047.0  United Kingdom  
13  2010-12-01       1.65     13047.0  United Kingdom  
14  2010-12-01       4.25     13047.0  United Kingdom  
15  2010-12-01       4.95     13047.0  United Kingdom  
16  2010-12-01       9.95     13047.0  United Kingdom  
17  2010-12-01       5.95     13047.0  United Kingdom  
18  2010-12-01       5.95     13047.0  United Kingdom  
19  2010-12-01       7.95     13047.0  United Kingdom

# checking total Null values in whole data set

In [73]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  541909 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   541909 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB
